README.md
# OmniToken

## Introduction
OmniToken is a handy tool for getting OAuth tokens from multiple providers. It brings up a web page interface which lists the popular "Login with XXX" social buttons and displays the token and other basic information after user finishes the authentication flow.

It is a tiny Sinatra-based web application.

It heavily relies on [OmniAuth](https://github.com/intridea/omniauth) and its strategies. Theoretically, any OAuth provider can be supported as long as there is a OmniAuth strategy for it. It can be easily achieved by installing the startegy gem and adding a config file.

## Get Started
To install the latest stable version, run:

`gem install omnitoken`

After the gem and its dependencies are installed, run:

`omnitoken -s`

A web server is started. Visit `http://localhost:4567` in the web browser and you will see a page with the "No OAuth providers are configured." message.

Follow next section to add some providers. You need to restart the web server after adding new providers.

## Add Providers
To get the token from a provider, a tiny configuration file is required. OmniToken has been tested with following providers and a configuration template is shipped for each of them:
* Twitter
* Facebook
* Pocket

### Twitter
First, install the twitter strategy gem:

`gem install omniauth-twitter`

Copy the twitter template by running:

`omnitoken -a twitter`

A `twitter.yml` file will be created in `providers` folder in your working directory. Open it using your favorite editor and fill the fileds. The values can be found on your twitter application page you registered beforehand.

### Facebook
First, install the facebook strategy gem:

`gem install omniauth-facebook`

Copy the facebook template by running:

`omnitoken -a facebook`

A `facebook.yml` file will be created in `providers` folder in your working directory. Open it using your favorite editor and fill the fileds. The values can be found on your facebook application page you registered beforehand.

### Pocket
First, install the facebook strategy gem:

`gem install omniauth-pocket`

Copy the pocket template by running:

`omnitoken -a pocket`

A `pocket.yml` file will be created in `providers` folder in your working directory. Open it using your favorite editor and fill the fileds. The values can be found on your pocket application page you registered beforehand.

### Others
Theoretically, OmniToken can be support any provider as long as there is a OmniAuth strategy for it. Please follow these steps to add a provider which doesn't have a shipped template.

First, find and install the strategy gem for your provider. Assume it is `omniauth-example`.

`gem install omniauth-example`

Create a file named `example.ym` and put it in `providers` folder under your working directory. **NOTE: all providers configuration file should be put in `providers`.**

Refer to startegies' documentation and get the required argumements. Add a corresponding value to each argument. For example, the arguments are `foo` and `bar`. The `example.yml` should look like this:

```yaml
foo: "xxx"
bar: "xxx"
```

## Supported Ruby Versions
OmniToken is tested under 2.0.0.

## License
Copyright (c) 2013-2014 Minjie Zha. See [LICENSE](https://github.com/magic003/omnitoken/blob/master/LICENSE-MIT) for details.


CHANGELOG.md
# OmniToken Changelog

## v0.0.3
* Add Pocket support.

## v0.0.2
* Just bump the version, because I made a mistake when pushing the gem.

## v0.0.1
* First alphaish release as a gem.
* Twitter and Facebook are supported.


test/helper.rb
ENV['RACK_ENV'] = 'test'
require 'minitest/autorun'
require 'rack/test'

begin
  require 'omnitoken_app'
rescue LoadError
  require File.join(File.expand_path(File.dirname(__FILE__)), %w{.. omnitoken_app})
end



test/omnitoken_app_test.rb
require File.expand_path(File.dirname(__FILE__)) + '/helper'

include Rack::Test::Methods

def app
  OmniTokenApp
end

describe 'OmniToken app' do
  it "should show the home page" do
    get '/'
    assert_equal 200, last_response.status
  end
end


app.rb
require 'sinatra/base'
require 'haml'
require 'omniauth'

require 'yaml'

class OmniTokenApp < Sinatra::Base
  PROVIDER_DIR = 'providers'

  enable :sessions
  set :haml, {:format => :html5, :layout => :layout}

  before '/*' do
    load_providers() unless loaded?
  end

  get '/' do
    haml :index
  end

  get "#{OmniAuth.config.path_prefix}/:provider/callback" do |p|
    auth = env['omniauth.auth']
    provider = get_provider(p)
    not_found if provider.nil?
    args = {}
    provider[:klass].args.each do |arg|
      args[arg] = provider[:args].shift
    end
    haml :results, :locals => { auth: env['omniauth.auth'], args: args}
  end

  get "#{OmniAuth.config.path_prefix}/failure" do
    [400, haml(:error, :locals => { message: params[:message] })]
  end

  not_found do
    [404, haml(:error, :locals => { message: "Page Not Found" })]
  end


  private

  def loaded?
    not (@providers.nil? || @providers.empty?)
  end

  def get_provider(name)
    provider = nil
    @providers.each do |p|
      provider = p if p[:id].eql?(name.to_sym)
    end
    provider
  end

  def load_providers
    @providers = []
    Dir.glob(File.join(PROVIDER_DIR, '*.yml')) do |file|
      name = File.basename(file, '.yml').to_sym
      strategy = require_strategy(name)
      @providers << { id: name,
                      display_name: OmniAuth::Utils.camelize(name),
                      path: "#{OmniAuth.config.path_prefix}/#{name}",
                      args: read_strategy_arg(strategy, file),
                      klass: strategy
                    }
    end

    use_strategy_middleware
  end

  def use_strategy_middleware
    providers = @providers
    self.class.use OmniAuth::Builder do
      providers.each do |p|
        provider p[:id], *p[:args]
      end
    end
  end

  def read_strategy_arg(strategy, file) 
    args = []
    yaml = YAML::load(File.open(file))
    strategy.args.each do |arg|
      if yaml[arg.to_s].nil?
        fail "The #{arg} argument is not provided in #{file}."
      else
        args << yaml[arg.to_s]
      end
    end
    args
  end

  def require_strategy(name)
    begin
      require "omniauth-#{name}"
      OmniAuth::Strategies.const_get(OmniAuth::Utils.camelize(name))
    rescue LoadError => e
      fail e, "Could not find matching strategy for #{name}. You may need to install an additional gem (such as omniauth-#{name})."
    end
  end

  run! if app_file == $0

end


omnitoken_app.rb
require 'sinatra/base'
require 'haml'
require 'omniauth'

require 'yaml'

# This is the Sinatra application for getting tokens.
class OmniTokenApp < Sinatra::Base
  # version number 
  VERSION = '0.0.3'

  # configuration folder for various OAuth providers
  PROVIDER_DIR = 'providers'

  enable :sessions
  set :haml, {:format => :html5, :layout => :layout}

  before '/*' do
    # load providers and save in the application level
    if !settings.respond_to?(:providers) || settings.providers.nil?
      load_providers()
      self.class.set :providers, @providers
    else
      @providers = settings.providers
    end
  end

  # home page
  get '/' do
    haml :index
  end

  # callback route for providers
  get "#{OmniAuth.config.path_prefix}/:provider/callback" do |p|
    auth = env['omniauth.auth']
    provider = get_provider(p)
    not_found if provider.nil?
    # get the arguments and provided values for the provider
    args = {}
    p_args = provider[:args].dup
    provider[:klass].args.each do |arg|
      args[arg] = p_args.shift
    end
    haml :results, :locals => { auth: env['omniauth.auth'], args: args}
  end

  # OAuth failure route
  get "#{OmniAuth.config.path_prefix}/failure" do
    [400, haml(:error, :locals => { message: params[:message] })]
  end

  not_found do
    [404, haml(:error, :locals => { message: "Page Not Found" })]
  end


  private

  # Gets the loaded provider by name
  def get_provider(name)
    provider = nil
    @providers.each do |p|
      provider = p if p[:id].eql?(name.to_sym)
    end
    provider
  end

  # Loads providers from the configuration files
  def load_providers
    @providers = []
    Dir.glob(File.join(PROVIDER_DIR, '*.yml')) do |file|
      name = File.basename(file, '.yml').to_sym
      # require ruby files for the strategy
      strategy = require_strategy(name)
      @providers << { id: name,
                      display_name: OmniAuth::Utils.camelize(name),
                      path: "#{OmniAuth.config.path_prefix}/#{name}",
                      args: read_strategy_arg(strategy, file),
                      klass: strategy
                    }
    end

    # set the strategies as a middleware
    use_strategy_middleware
  end

  # Sets the OmniAuth builder as a middleware
  def use_strategy_middleware
    providers = @providers
    self.class.use OmniAuth::Builder do
      providers.each do |p|
        provider p[:id], *p[:args]
      end
    end
  end

  # Reads the arguments for a strategy from the configuration file.
  def read_strategy_arg(strategy, file) 
    args = []
    yaml = YAML::load(File.open(file))
    strategy.args.each do |arg|
      if yaml[arg.to_s].nil?
        fail "The #{arg} argument is not provided in #{file}."
      else
        args << yaml[arg.to_s]
      end
    end
    args
  end

  # Requires the ruby file for a strategy.
  def require_strategy(name)
    begin
      require "omniauth-#{name}"
      OmniAuth::Strategies.const_get(OmniAuth::Utils.camelize(name))
    rescue LoadError => e
      fail e, "Could not find matching strategy for #{name}. You may need to install an additional gem (such as omniauth-#{name})."
    end
  end

  # Run this application
  run! if app_file == $0

end
